// ts-generic-collections-linq/esm2015/lib/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 * @template T
 */
export function IEnumerable() { }
if (false) {
    /** @type {?} */
    IEnumerable.prototype.length;
    /**
     * @param {?} index
     * @return {?}
     */
    IEnumerable.prototype.elementAt = function (index) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.any = function (predicate) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.all = function (predicate) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.single = function (predicate) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.first = function (predicate) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.last = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.singleOrDefault = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.firstOrDefault = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.lastOrDefault = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.where = function (predicate) { };
    /**
     * @template TResult
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.select = function (predicate) { };
    /**
     * @template TOuter, TMatch, TResult
     * @param {?} outer
     * @param {?} conditionInner
     * @param {?} conditionOuter
     * @param {?} select
     * @param {?=} leftJoin
     * @return {?}
     */
    IEnumerable.prototype.join = function (outer, conditionInner, conditionOuter, select, leftJoin) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.groupBy = function (predicate) { };
    /**
     * @param {?} comparer
     * @return {?}
     */
    IEnumerable.prototype.orderBy = function (comparer) { };
    /**
     * @param {?} comparer
     * @return {?}
     */
    IEnumerable.prototype.distinct = function (comparer) { };
    /**
     * @param {?} list
     * @return {?}
     */
    IEnumerable.prototype.union = function (list) { };
    /**
     * @param {?} no
     * @return {?}
     */
    IEnumerable.prototype.skip = function (no) { };
    /**
     * @param {?} no
     * @return {?}
     */
    IEnumerable.prototype.take = function (no) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.sum = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.avg = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.min = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.max = function (predicate) { };
    /**
     * @param {?=} predicate
     * @return {?}
     */
    IEnumerable.prototype.count = function (predicate) { };
    /**
     * @param {?} predicate
     * @return {?}
     */
    IEnumerable.prototype.forEach = function (predicate) { };
    /**
     * @return {?}
     */
    IEnumerable.prototype.toArray = function () { };
    /**
     * @return {?}
     */
    IEnumerable.prototype.asEnumerable = function () { };
}
/**
 * @record
 * @template T
 */
export function IGroup() { }
if (false) {
    /** @type {?} */
    IGroup.prototype.groups;
    /** @type {?} */
    IGroup.prototype.list;
}
/**
 * @record
 * @template T
 */
export function IComparer() { }
if (false) {
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    IComparer.prototype.compare = function (x, y) { };
}
/**
 * @record
 * @template T
 */
export function IEqualityComparer() { }
if (false) {
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    IEqualityComparer.prototype.equals = function (x, y) { };
}
