// ts-generic-collections-linq/esm5/lib/common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { List } from './list';
/**
 * @template T
 */
var /**
 * @template T
 */
Group = /** @class */ (function () {
    function Group(groups, list) {
        this.list = new List();
        this.groups = groups;
        this.list = new List(list);
    }
    return Group;
}());
/**
 * @template T
 */
export { Group };
if (false) {
    /** @type {?} */
    Group.prototype.groups;
    /** @type {?} */
    Group.prototype.list;
}
/** @type {?} */
export var objCompare = (/**
 * @param {?} obj1
 * @param {?} obj2
 * @return {?}
 */
function (obj1, obj2) {
    //Loop through properties in object 1
    for (var p in obj1) {
        //Check property exists on both objects
        if (obj1.hasOwnProperty(p) !== obj2.hasOwnProperty(p))
            return false;
        switch (typeof (obj1[p])) {
            //Deep compare objects
            case 'object':
                if (!objCompare(obj1[p], obj2[p]))
                    return false;
                break;
            //Compare function code
            case 'function':
                if (typeof (obj2[p]) == 'undefined' || (p != 'compare' && obj1[p].toString() != obj2[p].toString()))
                    return false;
                break;
            //Compare values
            default:
                if (obj1[p] != obj2[p])
                    return false;
        }
    }
    //Check object 2 for any extra properties
    for (var p in obj2) {
        if (typeof (obj1[p]) == 'undefined')
            return false;
    }
    return true;
});
/** @type {?} */
export var ITEM_NOT_FOUND_MSG = "Item does not exist.";
/** @type {?} */
export var MULTIPLE_INSTANCES_FOUND_MSG = "Multiple instances of entity found.";
