// rainbow-node-sdk/lib/common/models/ConferencePhoneNumber.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConferencePhoneNumber = void 0;
/// <summary>
/// Describe a phone number used to reach an audio Meeting
/// </summary>
class ConferencePhoneNumber {
    constructor() {
    }
    get location() {
        return this._location;
    }
    set location(value) {
        this._location = value;
    }
    get locationcode() {
        return this._locationcode;
    }
    set locationcode(value) {
        this._locationcode = value;
    }
    get number() {
        return this._number;
    }
    set number(value) {
        this._number = value;
    }
    get needLanguageSelection() {
        return this._needLanguageSelection;
    }
    set needLanguageSelection(value) {
        this._needLanguageSelection = value;
    }
    get numberType() {
        return this._numberType;
    }
    set numberType(value) {
        this._numberType = value;
    }
    /// <summary>
    /// Serialize this object to String
    /// </summary>
    /// <returns><see cref="String"/> as serialization result</returns>
    ToString() {
        return "";
        //return $"Location:[{Location}] - Locationcode:[{Locationcode}] - Number:[{Number}] - NumberType:[{NumberType}] - NeedLanguageSelection:[{NeedLanguageSelection}]";
    }
}
exports.ConferencePhoneNumber = ConferencePhoneNumber;
module.exports.ConferencePhoneNumber = ConferencePhoneNumber;
