// rainbow-node-sdk/lib/common/models/Settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settings = void 0;
const PresenceRainbow_1 = require("./PresenceRainbow");
/**
 * Enum for Presence values.
 * @readonly
 * @enum {string}
 */
/*let RainbowPresence = {
    ONLINE: "online",
    AWAY: "away",
    INVISIBLE: "invisible",
    DND: "dnd"
}; // */
/**
 * @class
 * @name Settings
 * @description
 *      This class represents a Settings. <br>
 *		Settings contains several informations stored and shared by application clients.<br>
 */
class Settings {
    constructor() {
        /**
         * @public
         * @readonly
         * @property {PresenceRainbow} presence Setting for manual user presence (used to go back to this presence when user logs in, instead of default (online))
         * @instance
         */
        this.presence = new PresenceRainbow_1.PresenceRainbow(PresenceRainbow_1.PresenceLevel.Online);
        /**
         * @public
         * @readonly
         * @property {boolean} displayNameOrderFirstNameFirst Setting for user display name order
         *  true: firstname first
         *  false: lastname first
         * @instance
         */
        this.displayNameOrderFirstNameFirst = true;
        /**
         * @public
         * @readonly
         * @property {string} activeAlarm Setting for active user alarm sound
         * @instance
         */
        this.activeAlarm = "";
        /**
         * @public
         * @readonly
         * @property {string} activeNotif Setting for active user notification sound
         * @instance
         */
        this.activeNotif = "";
    }
}
exports.Settings = Settings;
module.exports = {
    Settings: Settings
};
